File: algorithms/10-RegularExpressionMatching/Solution.java
class Solution {
    public boolean isMatch(String s, String p) {
        boolean[][] dp = new boolean[s.length() + 1][p.length() + 1];
        dp[s.length()][p.length() ]=true;
        
        for (int i= s.length();i>=0 ;i-- ){
            for(int j =p.length()-1; j>=0;j--){
                boolean firstMatch=i <s.length() && ( p.charAt(j)== s.charAt(i) || p.charAt(j)=='.');
                if( p.length() - j >=2 && p.charAt(j+1)=='*'){
                   dp[i][j]= (firstMatch && dp[i+1][j]) || dp[i][j+2];
                }else{
                    dp[i][j] =firstMatch && dp[i+1][j+1];
                }
                              
            }
        }
        // for (boolean[] ds:dp){
        //     System.out.println(Arrays.toString(ds));
        // }
        return dp[0][0];
    }
}

File: algorithms/10-RegularExpressionMatching/Solution.py
class Solution:
    def isMatch(self, s: str, p: str) -> bool:
        mem={}
        def dp(i,j):
            if(i,j) not in mem:
                if not p[j:] :
                    mem[i,j]=not s[i:]   
                    return mem[i,j]
                first_match= bool(s[i:]) and p[j] in [s[i],'.','*']
                if len(p)-j >=2 and p[j+1]=="*":
                    mem[i,j] = (first_match and dp(i+1,j)) or dp(i,j+2) 
                else:
                    mem[i,j]= first_match and dp(i+1,j+1)
                
            return mem[i,j]
                
        return dp(0,0)
    
    def isMatch2(self, s: str, p: str) -> bool:
        if not p :return not s    
        first_match= bool(s) and p[0] in [s[0],'.','*']
        if len(p)>=2 and p[1]=="*":
            return (first_match and self.isMatch(s[1:],p)) or self.isMatch(s,p[2:])         
        else:
            # first_match= bool(s) and p[0] in [s[0],'.','*']
            return first_match and self.isMatch(s[1:],p[1:])
            
     
